In [ ]:
import yaml

with open('../config.yaml', 'r') as f:
    cfg = yaml.safe_load(f)

# Numpy

* Пакет для Python, реализующий многомерные массивы
* Массивы **статически типизированные**. Тип элементов определяется при создании массива.
* Массивы эффективно используют память.
* Методы линейной алгебры в numpy реализованы на C и Fortran, что обеспечивает хорошую производительность.
* Пакет numpy активно используется в научных проектах по всему миру.

In [1]:
import numpy as np

## Создание массивов

Существует несколько способов создания новых numpy массивов:
* через списки и кортежи Python
* используя специальные функции numpy такие, как `arange`, `linspace`, и так далее.
* вычитывая данные из файлов

### 1D массивы:

In [2]:
a = np.array([0,1,2,3])
a

array([0, 1, 2, 3])

In [3]:
type(a), a.dtype

(numpy.ndarray, dtype('int64'))

In [4]:
a.ndim, a.shape, len(a)

(1, (4,), 4)

In [5]:
b = np.array((3, 4, 5))
b

array([3, 4, 5])

В numpy существует множество функций для генерации массивов:

<p style="font-size:16px"><b>`arange`</p></b>

Генерирует значения в интервале [start, stop) с шагом step. Аналог встроенной функции Python `range`. На уровне типов они, конечно, отличаются. `np.arange` возвращает обычный numpy-массив, в то время как `range` является lazy sequence/lazy iterable и стоит в общем ряду с `list` и `tuple`. Funny fact: `range` не является итератором -- он не "иссякает", для него нельзя вызвать `next()` и можно `len()`, но, в отличие от списка, он lazy

In [6]:
x = np.arange(0, 10, 1) # аргументы: start, stop, step
type(range(10))

range

In [7]:
x = np.arange(-1, 1, 0.1)
x

array([-1.00000000e+00, -9.00000000e-01, -8.00000000e-01, -7.00000000e-01,
       -6.00000000e-01, -5.00000000e-01, -4.00000000e-01, -3.00000000e-01,
       -2.00000000e-01, -1.00000000e-01, -2.22044605e-16,  1.00000000e-01,
        2.00000000e-01,  3.00000000e-01,  4.00000000e-01,  5.00000000e-01,
        6.00000000e-01,  7.00000000e-01,  8.00000000e-01,  9.00000000e-01])

<p style="font-size:16px"><b>`linspace` и `logspace`</b></p>

`linspace` Генерирует равномерно распределенные числа, включая конечные точки.

`logspace` То же, но в логарифмической шкале.

In [10]:
np.linspace(0, 10, 10) # аргументы: start, stop, число точек

array([ 0.        ,  1.11111111,  2.22222222,  3.33333333,  4.44444444,
        5.55555556,  6.66666667,  7.77777778,  8.88888889, 10.        ])

In [11]:
np.logspace(0.1, 1, 4, base=2)

array([1.07177346, 1.31950791, 1.62450479, 2.        ])

<p style="font-size:16px"><b>`zeros`, `ones`, `zeros_like` и `ones_like`</b></p>

In [12]:
np.zeros((5,))   # Аргумент должен быть кортежем

array([0., 0., 0., 0., 0.])

In [13]:
a = np.ones((4,))
a

array([1., 1., 1., 1.])

In [14]:
b = np.zeros_like(a)
b

array([0., 0., 0., 0.])

In [15]:
c = np.ones_like(b)
c

array([1., 1., 1., 1.])

Задание 1. Создайте вектор со значениями от 0 до 65

In [16]:
vector = np.arange(0, 66)
vector

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65])

### Многомерные массивы

In [17]:
# Матрица
m = np.array([[1., 2.], [3., 4.]])
m

array([[1., 2.],
       [3., 4.]])

In [18]:
type(m), m.dtype

(numpy.ndarray, dtype('float64'))

In [19]:
m.ndim, np.shape(m), len(m), np.size(m)

(2, (2, 2), 2, 4)

При попытке назначить значение другого типа будет выдана ошибка:

In [20]:
m[0,0] = 'hello' 

ValueError: could not convert string to float: 'hello'

<p style="font-size:16px"><b>`zeros`, `ones`, `zeros_like` и `ones_like`</b></p>

In [21]:
a = np.ones((3, 3))
a

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [22]:
b = np.zeros((2, 2))
b

array([[0., 0.],
       [0., 0.]])

<p style="font-size:16px"><b>Другие функции</b></p>

In [23]:
c = np.eye(3) # единичная матрица
c

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [24]:
d = np.diag(np.arange(4)) # диагональная матрица
d

array([[0, 0, 0, 0],
       [0, 1, 0, 0],
       [0, 0, 2, 0],
       [0, 0, 0, 3]])

**Задание**: Создайте матрицу 3х3 со значениями от 0 до 9

In [26]:
np.arange(9).reshape((3,3))

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [25]:
np.arange(9).reshape((3,3))

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

**Задание**: создайте матрицу 9 на 9, заполните случайными значениями и найдите максимальный, минимальный элемент и среднее значение.

In [ ]:
# Your code
# hint: если не помните, как делать ту или иную часть задачи, смотрите материалы дальше

## Копирование в NumPy

Как мы помним, в Python при присваивании не происходит копирование объектов. 

In [27]:
m = np.array([[1, 2], [3, 4]])
m

array([[1, 2],
       [3, 4]])

In [28]:
n = m

In [29]:
# Изменение N меняет M
n[0, 0] = 10
n

array([[10,  2],
       [ 3,  4]])

In [30]:
m

array([[10,  2],
       [ 3,  4]])

Глубокая копия создается в NumPy с помощью функции `copy`:

In [ ]:
n = np.copy(m)

In [ ]:
# теперь при изменении N M остается нетронутым
n[0,0] = -5
n

In [ ]:
m

Какой бы shape ни имел массив, в памяти он хранится как обыкновенный линейный массив. В примере показан C-order хранения многомерного массива ("укладываем" массив по строкам 
 последний индекс является самым быстрым). Такие массивы называют C-contiguous. Альтернативой является Fortran-order, где массив "укладывается" по столбцам 
 первый индекс является самым быстрым. Такие массивы называют Fortran-contiguous. Узнать, является ли массив C-contiguous или Fortran-contiguous, мы можем с помощью атрибута flags. В общем случае numpy используется C-order:

In [ ]:
a = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
])
a.flags

In [ ]:
a = np.array([
    [1, 4, 7],
    [2, 5, 8],
    [3, 6, 9],
])
a.flags

Линейный массив (он же иногда называется буфером), в котором хранятся данные, хранится в атрибуте data. Более конкретно, в нем хранится указатель на первый элемент массива, поэтому работать с ним напрямую мы не можем:

In [ ]:
a.data

Атрибут strides задает шаги в байтах, необходимые для перехода к следующему элементу вдоль того или иного индекса. Например, для двумерного массива типа int16 с shape = (4, 3) и C-contiguous буфером мы получим stride = (3x2, 1x2)

Numpy старается минимизировать количество копирований массивов, в результате чего многие операции, например slicing и в большинстве случаев reshape, возвращают view (представление) массива, а не его копию. Это становится возможным благодаря модификации shape, strides и указателя на первый элемент массива (то есть data). Помня об этом, всегда можно догадаться, произойдет копирование или будет использоваться view: всегда, когда можно построить массив из исходного путем модификации shape, stride и/или сдвига указателя в data, будет использоваться view. Рассмотрим несколько примеров

In [ ]:
a = np.array([
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [9, 10, 11],
], dtype=np.int16)
a.strides

Воспользуемся slicing для взятия каждого второго столбца. Это эквивалентно удвоению stride по второму индексу. Следовательно, будет использоваться view без копирования массива

In [ ]:
b = a[:, ::2]
b.strides

Проверить, является ли массив view, можно, обратившись к атрибуту base: он равен None для скопированных или вновь созданных массивов и равен исходному массиву для view

In [ ]:
b.base

In [ ]:
b.copy().base is None

Аналогично можно убедиться, что slicing для взятие каждой второй строки, начиная со второй, так же даст view без какого-либо копирования. В данном случае происходит удвоение stride по первому индексу и сдвиг указателя вперед на один элемент

In [ ]:
b = a[1::2, :]
b.strides

In [ ]:
b.base

Из этих примеров становится ясно, что все операции slicing-а приводят к появлению view. Более сложная индексация, которую мы рассмотрим ниже, уже приведет к копированию


При изменении представления меняется и изначальный массив:

In [ ]:
a = np.arange(10)
a

In [ ]:
b = a[::2]
b

In [ ]:
b[0] = 12
b

In [ ]:
a # (!!)

In [ ]:
a = np.arange(10)
b = a[::2].copy() # глубокое копирование
b[0] = 12
a

## Слияние массивов

Функции `vstack`, `hstack` и `concatenate` позволяются составить общий массив из нескольких массивов:

<p style="font-size:22px"><b>`concatenate`</b></p>

In [31]:
a = np.array([[1, 2], [3, 4]])
a

array([[1, 2],
       [3, 4]])

In [32]:
b = np.array([[5, 6]])
b

array([[5, 6]])

In [33]:
np.concatenate((a, b), axis=0)

array([[1, 2],
       [3, 4],
       [5, 6]])

In [34]:
np.concatenate((a, b.T), axis=1)

array([[1, 2, 5],
       [3, 4, 6]])

<p style="font-size:22px"><b>`hstack` и `vstack`</b></p>

In [35]:
np.vstack((a,b))

array([[1, 2],
       [3, 4],
       [5, 6]])

In [36]:
np.hstack((a,b.T))

array([[1, 2, 5],
       [3, 4, 6]])

## Индексация

Доступ к данным массива организуется с помощью индексов и оператора `[]`.

In [ ]:
a = np.arange(10)
a

In [ ]:
a[0], a[2], a[-1]

Для многомерных массивов индексами является кортеж целых чисел:

In [37]:
m = np.diag(np.arange(3))
m

array([[0, 0, 0],
       [0, 1, 0],
       [0, 0, 2]])

In [38]:
m[1, 1]

np.int64(1)

In [41]:
m[(1, 1)]

np.int64(1)

Можно использовать "`:`" для получения доступа к целой колонке или строке: 

In [39]:
m[1, :] # строка 1

array([0, 1, 0])

In [40]:
m[:, 2] # колонка 2

array([0, 0, 2])

Присваивание новых значений элементам массива:

In [ ]:
m[2, 1] = 10
m

In [ ]:
m[1, :] = 5
m[: ,2] = -1

In [ ]:
m

## Слайсинг

NumPy поддерживает слайсинг, как и списки с кортежами в Python:

In [42]:
a = np.arange(10)
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Все три параметра слайсинга являются опциональными: по умолчанию `start` равен **0**, `end` равен последнему элемену и `step` равен **1** в `a[start:stop:step]`:

In [43]:
a[::] # все параметры слайсинга имеют значения по умолчанию

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [44]:
a[1:3]

array([1, 2])

In [ ]:
a[:3]

In [ ]:
a[3:]

In [ ]:
a[2:9:2] 

Отрицательные индексы отсчитываются от конца массива:

In [ ]:
a[-1] # последний элемент массива

In [ ]:
a[-3:] # последние три элемента

Слайсы являются представлениями массива, а потому являются изменяемыми:

In [48]:
a[1:3] = [-2,-3]
a

array([ 0, -2, -3,  3,  4,  5,  6,  7,  8,  9])

**Задание**: разверните массив (1, 2, 3) -> (3, 2, 1)

In [45]:
b = np.arange(1, 4)
b[::-1]

array([3, 2, 1])

Слайсинг работает точно так же и для многомерных массивов:

In [46]:
m = np.random.randint(1,100, size=(4, 4))
m

array([[85, 10, 62,  6],
       [19, 39, 82, 32],
       [28, 28,  1, 12],
       [80, 98, 68, 55]], dtype=int32)

In [47]:
m[1:4, 1:4]

array([[39, 82, 32],
       [28,  1, 12],
       [98, 68, 55]], dtype=int32)

In [ ]:
m[::2, ::2]

**Задание**: Создайте массив 5х5 и заполните двойками обе диагонали, а нулями все остальное

In [53]:
a = np.diag(np.array([2, 2, 2, 2, 2]))
a = a[::-1]
np.fill_diagonal(a, 2)
a

array([[2, 0, 0, 0, 2],
       [0, 2, 0, 2, 0],
       [0, 0, 2, 0, 0],
       [0, 2, 0, 2, 0],
       [2, 0, 0, 0, 2]])

**Задание**: Создайте массив с значениями 1,2,3,4 на 1 поддиагонали 

In [61]:
a = np.diag(np.array([1, 2, 3, 4]), k=-1)
a

array([[0, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [0, 0, 3, 0, 0],
       [0, 0, 0, 4, 0]])

## Линейная алгебра

Код, написанный на numpy становится эффективным тогда, когда он достаточно векторизован (т.е. векторные и матричные операции используются в бОльшей части программы).

### Поэлементные операции

Все арифметические операции по умолчанию являются поэлементными:

In [62]:
a = np.arange(4)
a + 1

array([1, 2, 3, 4])

In [63]:
5*a

array([ 0,  5, 10, 15])

In [64]:
2**a

array([1, 2, 4, 8])

In [65]:
j = np.arange(5)
2**(j + 1) - j

array([ 2,  3,  6, 13, 28])

### Операции между массивами

In [66]:
a = np.arange(4)
b = np.ones(4) + 3
print('a = ', a)
print('b = ', b)
a - b

a =  [0 1 2 3]
b =  [4. 4. 4. 4.]


array([-4., -3., -2., -1.])

In [67]:
a * b

array([ 0.,  4.,  8., 12.])

Сравнения:

In [68]:
a = np.array([1, 2, 3, 4])
b = np.array([4, 2, 2, 4])
a == b

array([False,  True, False,  True])

In [69]:
a > b

array([False, False,  True, False])

Логические операции:

In [70]:
a = np.array([1, 1, 0, 0], dtype=bool)
b = np.array([1, 0, 1, 0], dtype=bool)
np.logical_or(a, b)

array([ True,  True,  True, False])

In [71]:
a | b

array([ True,  True,  True, False])

In [72]:
np.logical_and(a, b)

array([ True, False, False, False])

In [73]:
a & b

array([ True, False, False, False])

Использование логических операторов позволяет обращаться к элементам массива через маску:

In [74]:
c = np.array([1, 2, 3, 4], dtype=int)
d = np.array([1, 2, 5, 6], dtype=int)
c[c == d]

array([1, 2])

**Задание**: найдите ненулевые элементы массива [1,2,0,0,8,0, 4,1]

In [76]:
a = np.array([1,2,0,0,8,0, 4,1], dtype=int)
a[a != 0]

array([1, 2, 8, 4, 1])

### Матричная алгебра

Перемножение массивов является поэлементным. Для матричного умножения необходимо использовать функцию `dot`:

In [93]:
m = np.array([[1., 2.], [3., 4.]])
m

array([[1., 2.],
       [3., 4.]])

In [94]:
m * m

array([[ 1.,  4.],
       [ 9., 16.]])

In [95]:
m.dot(m)

array([[ 7., 10.],
       [15., 22.]])

**Задание**: Создайте 2 массива 2х4, проверьте, что они одинаковы

In [81]:
m = np.arange(8).reshape((2, 4))
b = np.arange(8).reshape((2, 4))
m == b

array([[ True,  True,  True,  True],
       [ True,  True,  True,  True]])

### Трансформирование массивов

Для транспонирования матриц используется либо `.T`, либо функция `transpose`:

In [ ]:
m

In [ ]:
m.T

In [ ]:
m.transpose()

Метод reshape позволяет изменить shape массива на произвольный shape, дающий то же число элементов, что и оригинальный shape. 

Понять, каким будет конечный вид массива, легко. Достаточно представить, как текущий массив выглядит в линейной форме в соответствии с C-order или F-order в зависимости от аргумента order, передаваемого в reshape (по умолчанию используется C-order), и переписать его в многомерный вид в соответствии с новым shape, используя тот же order. Рассмотрим пример ниже

In [82]:
a = np.array([
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [9, 10, 11],
], dtype=np.int16)
a.shape

(4, 3)

In [83]:
a.reshape(-1)


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int16)

Допустим теперь, что мы хотим получить shape = (2, 2, 3). В C-order самым быстрым индексом является последний, следовательно, фиксируя первые два индекса, мы должны получить подмассивы a_1 = [0, 1, 2], a_2 = [3, 4, 5], a_3 = [6, 7, 8] и a_4 = [9, 10, 11]. Их "схлопывание" дает снова линейный массив [a_1, a_2, a_3, a_4]. Снова используя C-order и фиксируя первый индекс, мы получаем два подмассива [a_1, a_2] и [a_3, a_4]. Следовательно, финальный массив будет иметь следующий вид:

[
    [
        [0, 1, 2],
        [3, 4, 5],
    ],
    [
        [6, 7, 8],
        [9, 10, 11],
    ],
]

In [84]:
a.reshape(2, 2, 3)

array([[[ 0,  1,  2],
        [ 3,  4,  5]],

       [[ 6,  7,  8],
        [ 9, 10, 11]]], dtype=int16)

**Задание**: Создайте вектор х длины 5 и матрицу А размера 5х4. Перемножьте их, чтобы получить результаты размера [1,5], [5], [4x5] 

In [127]:
x = np.arange(5)
A = np.arange(20).reshape((5, 4))
result_1 = np.dot(x, A)
result_2 = np.dot(A.T, x)
result_1, result_2

(array([120, 130, 140, 150]), array([120, 130, 140, 150]))

Другие математические функции:

In [114]:
c = np.matrix([[1j, 2j], [3j, 4j]])
c

matrix([[0.+1.j, 0.+2.j],
        [0.+3.j, 0.+4.j]])

In [ ]:
np.conjugate(c)

Эрмитово-сопряженная матрица(transpose + conjugate):

In [ ]:
c.H

Вещественная и мнимая части могут быть получены с помощью `real` и `imag`:

In [ ]:
np.real(c) # то же: c.real

In [ ]:
np.imag(c) # то же: c.imag

Модули элементов матрицы:

In [ ]:
np.abs(c)

Задание: Извлеките целые части из случайного массива размера 5х5 (с float типом) как минимум 3 разными способами

In [115]:
a = np.random.rand(5, 5) * 10
b = np.floor(a).astype(int)
b

array([[4, 2, 2, 0, 3],
       [3, 6, 4, 8, 6],
       [3, 0, 0, 8, 4],
       [4, 4, 9, 1, 7],
       [3, 8, 5, 8, 7]])

In [116]:
c = np.trunc(a).astype(int)
c

array([[4, 2, 2, 0, 3],
       [3, 6, 4, 8, 6],
       [3, 0, 0, 8, 4],
       [4, 4, 9, 1, 7],
       [3, 8, 5, 8, 7]])

In [117]:
d = (a // 1).astype(int)
d

array([[4, 2, 2, 0, 3],
       [3, 6, 4, 8, 6],
       [3, 0, 0, 8, 4],
       [4, 4, 9, 1, 7],
       [3, 8, 5, 8, 7]])

Задание: создайте двумерный массив, а потом добавьте к нему границу толщины 2 (весь исходный массив должен быть в новом)

In [119]:
a = np.random.rand(5, 5) * 10
b = np.pad(a, pad_width=2, mode='constant', constant_values=0)
b

array([[0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 6.59683995, 2.00990299, 5.97812423,
        7.82880536, 4.57786134, 0.        , 0.        ],
       [0.        , 0.        , 6.32273735, 5.83840647, 0.11242589,
        4.62419077, 1.67188908, 0.        , 0.        ],
       [0.        , 0.        , 8.58813071, 3.77158338, 4.16078268,
        8.78276294, 2.27377745, 0.        , 0.        ],
       [0.        , 0.        , 2.00273694, 8.17987583, 3.11959308,
        4.54768652, 4.19055695, 0.        , 0.        ],
       [0.        , 0.        , 8.34984516, 8.07967206, 1.50236554,
        3.68924751, 8.59084336, 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ,
        0.        , 0.        , 0.        , 0.        ],


Задание: создайте массив 8х8, заполните 0 и 999 в шахматном порядке. 

In [122]:
c = np.zeros((8, 8))

c[1::2, ::2] = 999
c[::2, 1::2] = 999

c

array([[  0., 999.,   0., 999.,   0., 999.,   0., 999.],
       [999.,   0., 999.,   0., 999.,   0., 999.,   0.],
       [  0., 999.,   0., 999.,   0., 999.,   0., 999.],
       [999.,   0., 999.,   0., 999.,   0., 999.,   0.],
       [  0., 999.,   0., 999.,   0., 999.,   0., 999.],
       [999.,   0., 999.,   0., 999.,   0., 999.,   0.],
       [  0., 999.,   0., 999.,   0., 999.,   0., 999.],
       [999.,   0., 999.,   0., 999.,   0., 999.,   0.]])

Задание: создайте массив 8х8, заполните 0 и 999 в шахматном порядке. Используйте функцию tile

In [124]:
b = np.array([[0, 999], [999, 0]])

c = np.tile(b, (4, 4))
c

array([[  0, 999,   0, 999,   0, 999,   0, 999],
       [999,   0, 999,   0, 999,   0, 999,   0],
       [  0, 999,   0, 999,   0, 999,   0, 999],
       [999,   0, 999,   0, 999,   0, 999,   0],
       [  0, 999,   0, 999,   0, 999,   0, 999],
       [999,   0, 999,   0, 999,   0, 999,   0],
       [  0, 999,   0, 999,   0, 999,   0, 999],
       [999,   0, 999,   0, 999,   0, 999,   0]])

### Матричные вычисления

<p style="font-size:18px"><b>inverse</b></p>

In [ ]:
np.linalg.inv(c) # то же: C.I 

In [ ]:
c.I * c

<p style="font-size:18px"><b>determinant</b></p>

In [ ]:
np.linalg.det(c)

In [ ]:
np.linalg.det(c.I)

Задание: нормализуйте случайный массив

In [128]:
random_array = np.random.rand(10) * 100

mean = np.mean(random_array)
std_dev = np.std(random_array)
normalized_array = (random_array - mean) / std_dev
random_array, normalized_array

(array([ 9.98694759,  6.07673675, 92.35154022, 17.32064829, 48.51634935,
        54.39048502, 15.82720354, 53.45435092, 16.767876  , 23.14913739]),
 array([-0.9133274 , -1.06339991,  2.2477967 , -0.63186254,  0.56541749,
         0.79086475, -0.68918043,  0.75493626, -0.65307775, -0.40816718]))

Задание: Создайте квадратные матрицы A, B. Посчитайте ((A+B)*(-A/2)) без копирования массивов.

In [129]:
n = 3
A = np.random.rand(n, n)
B = np.random.rand(n, n)
result = (A + B) * (-A / 2)
A, B, result


(array([[0.02146569, 0.22756039, 0.29330809],
        [0.43079949, 0.10178875, 0.4175737 ],
        [0.80590586, 0.97252935, 0.89244084]]),
 array([[0.09846746, 0.46214287, 0.32908874],
        [0.01294911, 0.37443648, 0.96025277],
        [0.78441135, 0.02820222, 0.85389288]]),
 array([[-0.00128722, -0.07847457, -0.09127701],
        [-0.09558334, -0.02423719, -0.28767205],
        [-0.64082298, -0.48662042, -0.77924977]]))

## Векторизация функций

Numpy предлагает средства для создания векторизованных оберток над функциями, которые изначально принимают на вход скалярные значения.

In [ ]:
def foo(x):
    if x >= 0:
        return 1
    else:
        return 0

In [ ]:
foo(np.array([-3, -2, -1, 0, 1, 2, 3]))

Эта функция работает для скалярных данных. 

Чтобы это функция принимала векторные значения, необходимо провести векторизацию с помощью функии `vectorize`:

In [ ]:
foo_vec = np.vectorize(foo)

In [ ]:
foo_vec(np.array([-3, -2, -1, 0, 1, 2, 3]))

- Задание: создайте случайный вектор размера 10 и отсортируйте
- Доп.задание: отсортируйте строки двумерной матрицы 6х6

In [ ]:
# your code

Задание: Вычтите среднее из строк матрицы.

In [ ]:
# your code

Задание: Создайте массив [[1, 0, null, 2, 4, null, 9],[null, 1, 0, -2, null, null, 3]]. Выведите маску, в которой 1 соответствуют null значениям.

In [ ]:
# your code

# Matplotlib

* Пакет для Python, используемый для создания качественных 2D визуализацией (есть минимальная поддержка 3D)
* Возможность создавать интерактивные графики
* Добавление множества графиков на один рисунок с кастомным расположением
* Экспорт в различные форматы изображений
* Есть поддержка анимаций

In [ ]:
# Позволяет matplotlib отображать графики сразу в notebook.
%matplotlib inline

## Matplotlib API

Импортирование модуля `matplotlib.pyplot` под именем `plt`:

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
import numpy as np

Простейший пример построения графиков в matplotlib:

In [ ]:
x = np.linspace(-2., 2., 128, endpoint=True)
y1 = x**2
y2 = np.exp(x)
plt.plot(x, y1)
plt.plot(x, y2)
plt.xlabel(r'$x \in \mathbb{R}$', fontsize=12)
plt.ylabel(r'$y(x)$', fontsize=12)
plt.show()

Рекомендуется создавать отдельный объект для каждого более-менее сложного графика. Это можно реализоваться, например, с помощью функии `subplots`:

In [ ]:
# Сетка графиков -- 1x1. Размер задается с помощью figsize.
fig, ax = plt.subplots(nrows=1, ncols=1, figsize=(8, 6))
ax.plot(x, y1, color="blue", linewidth=1.0, linestyle="-")
ax.plot(x, y2, color="green", linewidth=1.0, linestyle="--")
ax.grid()
ax.set_xlabel(r'$x$', fontsize=12)
ax.set_ylabel(r'$y$', fontsize=12)
plt.show()

Множественные графики

In [ ]:
# Создаем 2 графика (в 2 колонках)
fig, axes = plt.subplots(1, 2, figsize=(10, 6))
axes[0].plot(x, y1, 'r')
axes[1].plot(x, y2, 'b')
fig.tight_layout()

## Сохранение графиков

Текущий график можно сохранить, вызвав метод `savefig` класса `Figure`:

In [ ]:
fig.savefig("filename.png")

Также можно указать DPI и различные форматы:

In [ ]:
fig.savefig("filename.pdf", dpi=200)

### Легенды, описания осей и графиков

**Заголовок графика**


`axes.set_title("title")`

**Описания осей**


`axes.set_xlabel("x")
axes.set_ylabel("y")`

**Легенда**

Легенды могут создаваться двумя способами. Первый -- явно через метод `legend`:

`axes.legend(["curve1", "curve2"])`

Второй метод -- использование `label="label text"` при вызове `plot` с последующим вызовом метода `legend`: 

`axes.plot(x, x**2, label="curve1")
axes.plot(x, x**3, label="curve2")
axes.legend()`

Также можно выбрать расположение легенды на графике:

`ax.legend(loc=0) # автовыбор
ax.legend(loc='upper right')
ax.legend(loc='upper left')
ax.legend(loc='lower left')
ax.legend(loc='lower right')`

Пример использования описанного выше:

In [ ]:
fig, ax = plt.subplots()
ax.plot(x, x**2, label="y = x**2")
ax.plot(x, x**3, label="y = x**3")
ax.legend(loc='upper left');
ax.set_xlabel('x')
ax.set_ylabel('y')
ax.set_title('title');

**Задание**:
Дан временной ряд дневной цены некоторого актива:

stock_prices = np.array([100, 102, 101, 105, 107, 110, 108, 109, 112, 115, 114, 116, 118, 120, 119])
1) Постройте временной ряд недельного скользящего среднего, то есть скользящего среднего с окном, равным одной неделе. При построении скользящего среднего ряда игнорируйте ту часть исходного временного ряда, где в окно попадает меньше одной недели.
2) Постройте линейный график исходной цены и цены, полученной с помощью скользящего среднего

In [ ]:
# your code

Интерфейс subplots позволяет создавать сетку из множественных графиков

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(8, 4))
axes[0].plot(x, y1, "r")  # в этот раз мы используем fmt для задания цвета
axes[1].plot(x, y2, "b")
for ax in axes:
    ax.set_xlabel(r'$x$', fontsize=12)
    ax.set_ylabel(r'$y(x)$', fontsize=12)
    ax.grid()
fig.tight_layout()
plt.show()

**Задание**: создайте график 2х1, постройте графики sin(x), e**(x). Используйте в графике с e**x логарифмическую шкалу (semilogx, semilogy или loglog).

# Pandas

* Пакет для Python, используемый для представления, манипуляций и анализа табличных данных 
* Строится вокруг эффективно реализованных структур `pd.DataFrame` (таблица) и `pd.Series` (колонка), которые позволяют быстро манипулировать большими датасетами (если они помещаются в памяти, в противном случае необходимо использовать `dask`)
* Поддерживает интерфейс, схожий с numpy массивами, но добавляет поддержку кастомовой табличной индексации
* Часто используется как структура датасета для ML моделей

In [ ]:
import pandas as pd

## Создание рядов и датафреймов

Основные манипуляции данными в pandas происходят через `pd.Series` (одномерный массив с кастомной индексацией) и `pd.DataFrame` (двумерный массив с кастомной индексацией и интерпретацией строк и столбцов). В рамках одного столбца (или одного `pd.Series`) тип данных постоянен.

### Создание `pd.Series`

Мы можем создать `pd.Series` напрямую из массива/списка. В этом случае будет создан автоматический индекс от 0 до N - 1, к которому мы всегда можем получить доступ через аттрибут `index`. Сам индекс является отдельным типом, который наследуется от `pd.Index` и представляет immutable последовательность, пригодную для индексации.

In [ ]:
a = [1, 3, 5]
s = pd.Series(a)
s

In [ ]:
s.index

Мы также можем задать свой собственный индекс, если он у нас имеется. Для удобства можно представлять себе индекс как первичный ключ в терминах SQL

In [ ]:
s = pd.Series(a, index=['x', 'y', 'z'])
s

In [ ]:
s.index

Создать `pd.Series` мы можем также из словаря, где ключи будут интерпретироваться как индексы.

In [ ]:
data = {
    'x': 1,
    'y': 3,
    'z': 5,
}
s = pd.Series(data)
s

### Создание `pd.DataFrame`

Мы можем представить `pd.DataFrame` как набор `pd.Series`, скрепленных общим индексом. Существует множество способов, как мы можем создать `pd.DataFrame`, но проще всего это сделать с помощью словаря.

In [ ]:
people = {
    'first_name': ['Anastasiya', 'Anton', 'Boris'],
    'second_name': ['Nikolskaya', 'Pershin', 'Yeltsin'],
    'sex': ['f', 'm', 'm'],
    'gpa': [4.8, 4.5, 3.5],
}
df = pd.DataFrame(people)
df

Мы также можем загрузить данные из файла. Чаще всего для этого используют csv и соответствующую функции `pd.read_csv()`

In [ ]:
df_csv = pd.read_csv(cfg['house_pricing']['train_dataset'])
df_csv.head()

## Доступ к данным в датафрейме

Используя индекс и название столбца, мы всегда можем получить доступ к любому набору данных в датафрейме. Общий интерфейс нам предоставляет оператор `[ ]`, примененный к `loc`. Первым аргументом здесь будет индекс, а вторым -- столбец:

In [ ]:
df.loc[1, 'first_name']

И первый, и второй аргументы могут последовательностями, что позволяет получить доступ к любому подмножеству датафрейма, который сам является датафреймом:

In [ ]:
df.loc[[0, 1], ['first_name', 'second_name']]

Есть несколько упрощений для этого синтаксиса. Начнем с доступа по индексу исключительно. Допустим, мы хотим получить строку с индексом 1:

In [ ]:
df.loc[1]

Обратите внимание, что возвращенный объект является инстансом `pd.Series` с индексом `['first_name', 'second_name', 'sex', 'gpa']`:

In [ ]:
print(type(df.loc[1]))
df.loc[1].index

Мы также можем получить `pd.Series` по одному из столбцов, используя оператор `[ ]` напрямую:

In [ ]:
df['first_name']

Снова можно передать последовательность названий столбцов в `[ ]` для получения подмножества столбцов:

In [ ]:
df[['first_name', 'second_name']]

Мы также можем комбинировать `loc()` и `[ ]` для получения доступа к любому подмножеству датафрейма (chained indexing). Это легитимно до тех пор, пока нашей задачей является доступ к данным, а не их изменение

In [ ]:
df[['first_name', 'second_name']].loc[[0, 1]]

## Изменение данных в датафрейме

Обратите внимание, что во всех операциях выше возвращался новый инстанс `pd.DataFrame` или `pd.Series`. Возникает естественный вопрос -- происходит в этом случае копирование данных или нет? На этот вопрос нет однозначного ответа -- в одном случае может произойти копирование данных, а в другом будет возвращено их представление. Существуют правила, которые конкретизируют эти случаи [1], но во всех операциях следует использовать наиболее консервативный подход, то есть явные копирования и изменения. Ниже мы рассмотрим частные случаи.

Правило для изменения данных в исходном датафрейме очень простое -- не должно быть никакого chained indexing [2]. В противном случае мы можете получить warning `SettingWithCopy`. Причина состоит в том, что pandas не имеет механизма для отслеживания "представлений в представлениях". После первой индексации будет возвращен датафрейм/ряд, относительно которого нет уверенности -- является ли он копией или представлением.

[1] https://stackoverflow.com/questions/23296282/what-rules-does-pandas-use-to-generate-a-view-vs-a-copy  
[2] https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

Предположим, что мы хотим изменить оценку в исходном датафрейме для индекса 1. Мы можем это сделать, избегая chained indexing, используя `loc`:

In [ ]:
df.loc[1, 'gpa'] = 4.6
df

Попытка сделать это через chained indexing может не сработать:

In [ ]:
df.loc[1]['gpa'] = 4.5
df

А может и сработать:

In [ ]:
df['gpa'][1] = 4.5
df

Именно из-за такой неоднозначности и стоит использовать наиболее явные способы изменения датафреймов.

Еще одним из явных способов является изменение всего столбца через `[ ]`:

In [ ]:
df['gpa'] = [5.0, 4.5, 3.2]
df

Если требуется создать копию датафрейма, то следует также создать копию явно: 

In [ ]:
df_copy = df.copy()
df_copy.loc[0, 'gpa'] = 4.9
df_copy

## Exploratory data analysis (EDA) 

### Общая информация

Методы `head()` и `tail()` выводят первые (последние соотв.) n строк. Оба метода принимают кол-во выводимых строк как единственный аргумент (по умолчанию 5).

In [ ]:
df_csv.head()

In [ ]:
df_csv.tail()

Не все столбцы здесь выведены. Их список мы можем получить, используя аттрибут `columns`:

In [ ]:
df_csv.columns

Базовую информацию о датафрейме (кол-во строк, кол-во столбцов, типы столбцов, кол-во not NaN значений) мы можем получить, используя метод `info()`:

In [ ]:
df_csv.info()

NaN значения регулярно встречаются при работе с реальными датасетами. Как минимум, часто полезно знать их кол-во. Для этого удобно использовать метод `isna()` или `isnull()` (алиас для `isna()`):

In [ ]:
df_csv.isna()

Просуммируя столбцы, мы получим кол-во NaN значений в каждом столбце:

In [ ]:
df_csv.isna().sum()

Интересно узнать, какие столбцы имеют больше всего NaN значений. Для этого воспользуемся сортировкой:

In [ ]:
df_csv.isna().sum().sort_values(ascending=False)

**Задание**: найдите максимальное число пустых значений в строке.

**Задание**: сколько строк вообще не имеют пустых значений? Сколько строк имеет не больше одного?

### Статистическая информация

Основные статистики по столбцам можно получить, вызвав метод `describe()`:

In [ ]:
df_csv.describe()

Допустим, что нам интересны статистики только для цены жилья и года постройки:

In [ ]:
df_csv[['SalePrice', 'YearBuilt']].describe()

Мы также можем настроить вывод нужных квантилей для лучшего описания распределений, используя ключевое слово `percentiles`:

In [ ]:
df_csv[['SalePrice', 'YearBuilt']].describe(percentiles=[0.1, 0.3, 0.5, 0.7, 0.9])

Те же статистики мы можем вычислить отдельно. Например, можно воспользоваться методами `mean()`, `median()`, `std()`

In [ ]:
df_csv.mean(numeric_only=True)

In [ ]:
df_csv.median(numeric_only=True)

In [ ]:
df_csv.std(numeric_only=True)

Эти статистики отлично подходят для численных столбцов, но множество столбцом в данном датасете представляют категориальные переменные. Для каждого такого столбца мы можем изучить распределение принимаемых значений, используя `value_counts()`:

In [ ]:
# TA = typical average, Gd = good, Ex = excellent, Fa = fair
df_csv['KitchenQual'].value_counts()

Мы можем использовать аргумент `normalize=True` для построения функции вероятностей:

In [ ]:
df_csv['KitchenQual'].value_counts(normalize=True)

Иногда нам необходимо провести анализ внутри подмножеств, полученных группировкой относительно некоторых столбцов. В этом случае полезным оказывается метод `groupby()`:

In [ ]:
df_csv.groupby('KitchenQual').mean()

Рассмотрев среднюю цену по группам, мы можем сделать очевидный вывод, что жилье с качественной кухней стоит сильно дороже:

In [ ]:
df_csv.groupby('KitchenQual')['SalePrice'].mean()

### Графики

Несмотря на то, что мы уже знакомы с matplotlib и могли бы воспользоваться им для построения графиков, pandas предоставляет встроенные инструменты, которые особенно удобно использовать в jupyter notebook во время EDA.

Например, построив гистрограмму, можно сразу предположить, что стоимость жилья имеет логнормальное распределение (или по крайней мере несимметричное распределение с тяжелым правым хвостом). При этом метод `plot.hist()` может использован только для числовых столбцов.

In [ ]:
df_csv['SalePrice'].plot.hist()

Используя два столбца и `plot.scatter()`, мы можем взглянуть на зависимость между ними:

In [ ]:
df_csv.plot.scatter(x='YearBuilt', y='OverallQual')

Для категориальных столбцов мы можем вывести распределение значений, используя `value_counts()` и `plot.bar()`. Здесь же мы продемонстрируем, как можно передавать дополнительные аргументы для изменения визуализации

In [ ]:
df_csv['KitchenQual'].value_counts().plot.bar(figsize=(12, 6))

**Задание**:
Найдете вме столбцы с  NaN значениями. Отсортируйте по увеличению числа NaN и постройте barplot.

Отличный инструмент для графиков  - seaborn. Это обертка над matplotlib, которая также предоставляет более широкий набор опций

In [ ]:
import seaborn as sns
from matplotlib import pyplot as plt
sns.histplot(df_csv['SalePrice'])

**Задание**: Постройте гистограмму, но с 20 бинами.

In [ ]:
# Your code

Конечно, гистограммы можно строить сразу для нескольких категорий. Посмотрим на цены и продажи в зависимости от квартала.

**Задание**: постройте гистограммы для кварталов (столбец MoSold).

In [ ]:
sns.histplot(x=df_csv['SalePrice'], #yout_code# , element="step")

**Задание**: Постройте scatterplot через seaborn (scatter) для нескольких пар столбцов из датасета

In [ ]:
sns.scatterplot(x=df_csv["LotFrontage"], y=df_csv["SalePrice"])

In [ ]:
df_csv.head()

Стиль графиков можно менять, например, поменяв палитру:
```
sns.set(style="whitegrid")
sns.set_palette(["#89CFF0", "#FFB6C1"])
```

Полезно строить графики и для категориальных данных. Например, будет полезно посмотреть на характеристики цены для разных LotShape

Пример такого графика: catplot. В зависимости от значения kind то, как точки будут располагаться на осях. Попробуйте не задавать его и задать "kind"
    

In [ ]:
sns.catplot(data=df_csv, x="LotShape", y="SalePrice", kind="swarm")

**Задание**: Постройте этот график, но добавив еще одно измерение с двумя категорими. Как это сделать? Тоже через hue.

Есть также и более продвинутые методы. Например, можно построить графики зависимости целевой переменной от других столбцов.

In [ ]:
# Выберем все числовые столбцы из датасета, кроме индекса и цены 
numeric_col = list(df_csv.select_dtypes(include=['number']).columns)
numeric_col.remove("Id")
numeric_col.remove("SalePrice")

num_plots = len(numeric_col)
num_columns = 3
num_rows = num_plots // num_columns + (1 if num_plots % num_columns > 0 else 0)

plt.figure(figsize=(16, 4 * num_rows))

for i, col in enumerate(numeric_col):
    plt.subplot(num_rows, num_columns, i + 1)
    sns.regplot(x=df_csv[col], y=df_csv["SalePrice"], scatter_kws={'color': '#89CFF0'}, line_kws={'color': '#FFB6C1'})
    plt.title(f"{col} vs SalePrice")
    plt.xlabel(col)
    plt.ylabel("SalePrice")
    
plt.tight_layout()
plt.show()

Также иногда полезный, а иногда перегруженный график - pairplot. Его можно сделать даже для всего датасета.

**Задание**: сделайте pairplot для нескольких столбцов датасета. Лучше не выбирать больше 5 за раз!